### PR TITLE
Introduce the concept of an abstract execution context

### PR DESCRIPTION
**Pull request by bdash**
This can be used by bindings that wish cached Realms to be keyed on something other than the thread they were created on, such as a JavaScript context or Python interpreter state.

Related to realm/realm-js#552, and will allow realm-js to easily fix realm/realm-js#473.

This should allow the thread ID changes from c34b3db85d86bf542fb57cd34f393efeac87b929 to be dropped from the `js` branch, making it possible for realm-js to use master directly.

### CODE REVIEW

File: src/execution_context_id.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    };
+
+public:
+
+    // Convert from the representation used by Realm::Config, where the absence of an
+    // explicit abstract execution context indicates that the current thread's identifier
+    // should be used.
+    AnyExecutionContextID(util::Optional<AbstractExecutionContextID> maybe_abstract_id)
+    {
+        if (maybe_abstract_id)
```
Comment (jpsim) [inline src/execution_context_id.hpp]: Take it or leave it.

```cpp
*this = AnyExecutionContextID(maybe_abstract_id ? *maybe_abstract_id : std::this_thread::get_id());
```

File: src/execution_context_id.hpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    bool operator==(const AnyExecutionContextID& other) const
+    {
+        return m_type == other.m_type && std::memcmp(&m_storage, &other.m_storage, sizeof(m_storage)) == 0;
+    }
+
+    bool operator!=(const AnyExecutionContextID& other) const
+    {
+        return !(*this == other);
```
Comment (jpsim) [inline src/execution_context_id.hpp]: How is this different than this?

```cpp
return *this != other;
```

File: src/execution_context_id.hpp (hunk 1 above)
Comment (bdash) [inline src/execution_context_id.hpp]: That won't compile as each branch of the ternary expression is required to have the same type.

File: src/execution_context_id.hpp (hunk 2 above)
Comment (bdash) [inline src/execution_context_id.hpp]: We're implementing `!=` so we cannot use it here.

Comment (jpsim) [inline src/execution_context_id.hpp]: lol

File: src/shared_realm.cpp [hunk 3]
```diff
@@ -429,9 +432,12 @@ static void check_read_write(Realm *realm)
 
 void Realm::verify_thread() const
 {
-    if (m_thread_id != std::this_thread::get_id()) {
+    if (!m_execution_context.template contains<std::thread::id>())
```
Comment (tgoyne) [inline src/shared_realm.cpp]: These two shouldn't need `template` since `m_execution_context` isn't a derived type.

**Comment by bdash** [comment]
@tgoyne, can you please take a look at the new changes in 31f87ab that are needed for Android?

**Pull request merged**